src/v/raft/consensus.cc

```diff
@@ -1057,9 +1057,31 @@ consensus::cancel_configuration_change(model::revision_id revision) {
       config());
     return interrupt_configuration_change(
              revision,
-             [revision](raft::group_configuration cfg) {
+             [this, revision](raft::group_configuration cfg) {
+                 /**
+                  * Optimization of cancelling configuration change
+                  *
+                  * When the raft group reconfiguration advanced beyond the
+                  * point where nodes from the old configuration are demoted to
+                  * learners there is no point of cancellation as the resulting
+                  * configuration would require greater quorum then the one
+                  * required to finish current configuration change.
+                  *
+                  */
+                 if (cfg.get_state() == raft::configuration_state::joint) {
+                     if (!cfg.old_config()->learners.empty()) {
```

Comment:
so this condition detects if we are in step 5 of reconfiguration. I wonder what will happen if we are in step 6? Will we still cancel?

My hunch is that the reliable condition for allowing the cancellation should look something like "are all nodes from A still voters?".

A related, question I'm wondering how the cancellation worked before, when we tried to cancel in step 6? AFAICS we don't add nodes in A-B back as learners in `cancel_update_in_joint_state`.

Reply:
I updated the description. In our implementation there is no step 6. i.e. when all voters to be removed are learners we move directly to simple configuration.